lib/SetAPI/util.py

```diff
@@ -4,17 +4,57 @@
 import os
 import re
 from installed_clients.DataFileUtilClient import DataFileUtil
+from typing import Any
 
+KBASE_UPA_REGEX = re.compile(r"^\d+/\d+(/\d+)?$")
+WS_NAME_OBJ_NAME_REGEX = re.compile(r"^([\w\.\-\|]+:)?[\w\.\-\|]+/[\w\.\-\|]+$")
 
-def check_reference(ref):
-    """
-    Returns True if ref looks like an actual object reference: xx/yy/zz or xx/yy
-    Returns False otherwise.
+
+def check_reference(ref: str | None) -> bool:
```

Comment:
I went on a bit of a diversion trying to work out how to accurately test for an invalid reference after one of my tests persistently failed. Unfortunately it was only after I'd written all this and done a load of testing that I realised that the test that was failing was not using this method. 🤦

Reply:
Agree completely. I consulted the WS documentation for the rules -- I didn't realise that you can have a single character (e.g. "a" or ".") as an object name.